infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/external/ExternalServerExposer.java

```diff
@@ -60,18 +66,68 @@ public void expose(
       String serviceName,
       ServicePort servicePort,
       Map<String, ServerConfig> externalServers) {
-    Ingress ingress = generateIngress(machineName, serviceName, servicePort, externalServers);
-    k8sEnv.getIngresses().put(ingress.getMetadata().getName(), ingress);
+    List<Ingress> ingresses =
+        generateIngresses(machineName, serviceName, servicePort, externalServers);
+    for (Ingress ingress : ingresses) {
+      k8sEnv.getIngresses().put(ingress.getMetadata().getName(), ingress);
+    }
   }
 
-  private Ingress generateIngress(
+  private List<Ingress> generateIngresses(
       String machineName,
       String serviceName,
       ServicePort servicePort,
       Map<String, ServerConfig> ingressesServers) {
 
+    // all the unique servers need their separate ingresses. all the other servers can get lumped
+    // under a common ingress corresponding to the server and port.
+    Map<String, ServerConfig> nonUniqueServers = new HashMap<>();
+
+    // let's go through the servers. Build the unique ingresses straight away and collect the
+    // servers to be put behind the common ingress.
+    List<Ingress> ingresses =
+        ingressesServers
+            .entrySet()
+            .stream()
+            .map(
+                e -> {
+                  ServerConfig serverConfig = e.getValue();
+
+                  if (!serverConfig.isUnique()) {
+                    nonUniqueServers.put(e.getKey(), serverConfig);
+                    return null;
+                  } else {
+                    return generateIngress(
+                        machineName,
+                        serviceName,
+                        e.getKey(),
+                        servicePort,
+                        ImmutableMap.of(e.getKey(), serverConfig));
+                  }
+                })
+            .filter(Objects::nonNull)
+            .collect(Collectors.toCollection(ArrayList::new));
+
+    // now generate the common ingress unconditionally, even if there are no servers
```

Comment:
> even if there are no servers

Really? Does not `if (!nonUniqueServers.isEmpty())` mean generate only if there are some servers? What I missed?

Reply:
Good catch. A remnant of an older approach.